Unmapped aggs should not run pipelines if they delegate reduction

polyfractal opened the pull request:
`UnmappedTerms`/`UnmappedSampler`/`UnmappedSigTerms` aggs try to delegate reduction to a sibling object that is not unmapped.  That delegated agg will run the reductions, and also reduce any pipeline aggs.  But because delegation comes before running pipelines, the `Unmapped*` agg _also_ tries to run pipeline aggs.

This causes the pipeline to run twice, and potentially double it's output in buckets which can create invalid JSON and break when converting to maps.

This fixes the issue by toggling a flag in UnmappedTerms if it delegated away reduction so that it knows not to run pipeline aggs either.

Closes #33514

elasticmachine commented:
Pinging @elastic/es-search-aggs

colings86 commented:
> Another easy fix would be sorting the aggs before reduction, so that we preferentially choose mapped aggs as the first reduction object.

What would happen if all the aggs were unmapped, would we end up having to do the same delegation we do now?

> We'd probably need the unmapped aggs to implement a simple interface so we could tell them apart from other aggs,

This could also be done by having a method on InternalAggregation that is something like `isMapped()` which an unmapped agg could return false for?

What would be great is if we could find a way to make it so Unmapped aggs are not special at all and so we don't have to avoid using them for reduce or being careful about double execution

polyfractal commented:
> What would happen if all the aggs were unmapped, would we end up having to do the same delegation we do now?

Same delegation process, but no need for the flag to track state.  If there are only unmapped aggs, the first one will get chosen to run.  Then it will look through all the aggs, see they are also all unmapped and then just return itself skipping reduction entirely.  Then it'll also run pipelines.

> This could also be done by having a method on InternalAggregation that is something like isMapped() which an unmapped agg could return false for?

++ that'd work too, and be cleaner.

polyfractal commented:
@colings86 Ok, reworked to use a sorting approach.  Has an overall smaller change to the codebase, at the small cost of a sort.   I also removed the delegation ability from unmapped aggs, since it's not needed anymore.  If they get called to lead the reduction they will just return themselves, because there are no other mapped aggs.

There is a bit of a danger here that the caller won't know to sort first, and won't know that unmapped aggs just return themselves. Right now there's only one place that calls this code so it's reasonably contained.  But we could make the unmapped aggs throw a `UnsupportedOperationException` to avoid accidents.

Thoughts?

polyfractal commented:
Jenkins, run gradle build tests

polyfractal commented:
Jenkins, run gradle build tests

polyfractal commented:
Huzzah, tests passing again.  Think this is ready for another look.

colings86 commented on server/src/test/java/org/elasticsearch/search/aggregations/pipeline/MaxBucketIT.java:
```diff
@@ -475,4 +486,56 @@ public void testNested() throws Exception {
+     *
+     * Applies to any pipeline agg, not just max.
+     */
+    public void testFieldGetsWrittenOutTwice() throws Exception {
```
Isn't this testing that the field does NOT get written twice?

polyfractal replied:
Ah, yes, the name should be adjusted.  Named it while it was failing, but that doesn't make sense now that it's fixed :)

polyfractal commented:
Jenkins, run gradle build tests
